src/bin/manix.rs

```diff
@@ -41,107 +95,72 @@ fn main() -> Result<()> {
-            .map_err(|e| anyhow::anyhow!(e))
-            .context("Failed to update Nixpkgs documentation cache")
-        {
-            eprintln!("{:?}", e);
-        } else {
-            aggregate_source.add_source(Box::new(nixpkgs_doc));
-        }
+        build_source_and_add(
+            OptionsDatabase::new(OptionsDatabaseType::HomeManager),
+            "Home Manager Options",
+            &options_hm_cache_path,
+            &mut aggregate_source,
+        );
+
+        build_source_and_add(
+            OptionsDatabase::new(OptionsDatabaseType::NixOS),
+            "NixOS Options",
+            &options_nixos_cache_path,
+            &mut aggregate_source,
+        );
+
+        build_source_and_add(
+            nixpkgs_tree_docsource::NixpkgsTreeDatabase::new(),
+            "Nixpkgs Tree",
+            &nixpkgs_tree_cache_path,
+            &mut aggregate_source,
+        );
+
+        build_source_and_add(
+            xml_docsource::XmlFuncDocDatabase::new(),
+            "Nixpkgs Documentation",
+            &nixpkgs_doc_cache_path,
+            &mut aggregate_source,
+        );
     } else {
-        match std::fs::read(&options_hm_cache_path)
-            .context("Failed to read the cache file for Home Manager")
-        {
-            Ok(cache_bin) => {
-                let options_db: OptionsDatabase = bincode::deserialize(&cache_bin)
-                    .context("Failed to deserialize Home Manager cache")?;
-                aggregate_source.add_source(Box::new(options_db));
-            }
-            Err(e) => eprintln!("{:?}", e),
-        }
-
-        match std::fs::read(&options_nixos_cache_path)
-            .context("Failed to read the cache file for NixOS")
-        {
-            Ok(cache_bin) => {
-                let options_db: OptionsDatabase = bincode::deserialize(&cache_bin)
-                    .context("Failed to deserialize NixOS cache")?;
-                aggregate_source.add_source(Box::new(options_db));
-            }
-            Err(e) => eprintln!("{:?}", e),
-        }
-
-        match nixpkgs_tree_docsource::NixpkgsTreeDatabase::load(&nixpkgs_tree_cache_path)
-            .context("Failed to read the cache file for Nixpkgs Tree")
-        {
-            Ok(tree) => aggregate_source.add_source(Box::new(tree)),
-            Err(e) => eprintln!("{:?}", e),
-        }
-
-        match xml_docsource::XmlFuncDocDatabase::load(&nixpkgs_doc_cache_path)
-            .context("Failed to read the cache file for Nixpkgs Documentation")
-        {
-            Ok(doc) => aggregate_source.add_source(Box::new(doc)),
-            Err(e) => eprintln!("{:?}", e),
-        }
+        load_source_and_add(
+            std::fs::read(&options_hm_cache_path).map(|c| OptionsDatabase::load(&c)),
+            "Home Manager Options",
+            &mut aggregate_source,
+        );
+
+        load_source_and_add(
+            std::fs::read(&options_nixos_cache_path).map(|c| OptionsDatabase::load(&c)),
```

Comment:
couldn't we use `and_then` here to flatten the error structure beforehand? the `Errors` type should be able to represent that possible io error, shouldn't it?